### PR TITLE
Move Get and Store config logic from collectorimpl to configstoreimpl

### PR DESCRIPTION
**Pull request by mackjmr**
### What does this PR do?
Addresses comment: https://github.com/DataDog/datadog-agent/pull/27115#discussion_r1664647169 and https://github.com/DataDog/datadog-agent/pull/27115#discussion_r1665604697.

### Describe how to test/QA your changes
Launch the converged agent. Then run curl -k https://localhost:7777/ and make sure that the user and enhanced config are correct based on the passed configuration, and also make sure that any sensitive fields are redacted with '[REDACTED]'

### CODE REVIEW

File: comp/otelcol/configstore/impl/configstore.go [hunk 1]
```diff
@@ -26,16 +28,45 @@ func NewConfigStore() (configstore.Component, error) {
 	return &configStoreImpl{}, nil
 }
 
-// AddProvidedConf stores the config into configStoreImpl.
-func (c *configStoreImpl) AddProvidedConf(config *otelcol.Config) {
+func (c *configStoreImpl) AddConfigs(providedCPS otelcol.ConfigProviderSettings, enhancedCPS otelcol.ConfigProviderSettings, factories otelcol.Factories) error {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
```
Comment (dineshg13) [inline comp/otelcol/configstore/impl/configstore.go:33]: Please use write lock

**Review by amenasria — APPROVED** [review]
LGTM for Agent Infra owned files !

**Review by pducolin — COMMENTED** [review]
Ok for files owned by @DataDog/agent-devx-loops , but I let @DataDog/agent-devx-infra as they are more familiar with this file

File: tasks/modules.py [hunk 2]
```diff
@@ -340,7 +340,13 @@ def generate_dummy_package(ctx, folder):
                         ctx.run(
                             "go mod edit -replace github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector=github.com/open-telemetry/opentelemetry-collector-contrib/connector/datadogconnector@v0.103.0"
                         )
-
+                    if (
+                        mod.import_path == "github.com/DataDog/datadog-agent/comp/otelcol/configstore/impl"
+                        or mod.import_path == "github.com/DataDog/datadog-agent/comp/otelcol/configstore/def"
+                    ):
+                        ctx.run("go mod edit -exclude github.com/knadh/koanf/maps@v0.1.1")
+                        ctx.run("go mod edit -exclude github.com/knadh/koanf/providers/confmap@v0.1.0")
```
Comment (dineshg13) [inline tasks/modules.py:348]: why do we need this ?

File: comp/otelcol/configstore/impl/configstore.go [hunk 3]
```diff
@@ -26,18 +28,44 @@ func NewConfigStore() (configstore.Component, error) {
 	return &configStoreImpl{}, nil
 }
 
-// AddProvidedConf stores the config into configStoreImpl.
-func (c *configStoreImpl) AddProvidedConf(config *otelcol.Config) {
-	c.mu.RLock()
-	defer c.mu.RUnlock()
+func (c *configStoreImpl) AddConfigs(providedCPS otelcol.ConfigProviderSettings, enhancedCPS otelcol.ConfigProviderSettings, factories otelcol.Factories) error {
```
Comment (dineshg13) [inline comp/otelcol/configstore/impl/configstore.go:31]: Can we add them as Fx deps. It seems odd to post construction

File: comp/otelcol/configstore/impl/configstore.go [hunk 4]
```diff
@@ -26,16 +28,45 @@ func NewConfigStore() (configstore.Component, error) {
...
 
 	c.provided = config
 }
 
-// AddEnhancedConf stores the config into configStoreImpl.
-func (c *configStoreImpl) AddEnhancedConf(config *otelcol.Config) {
+// addEnhancedConf stores the config into configStoreImpl.
+func (c *configStoreImpl) addEnhancedConf(config *otelcol.Config) {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
```
[the remark this answers is not in the document]
Comment (mackjmr) [reply comp/otelcol/configstore/impl/configstore.go]: seems like you were looking at outdated commit, should be good now: https://github.com/DataDog/datadog-agent/pull/27337/files#diff-073a5a96f7bc973099a739a1b183f433630b2d4579e71d03393bfba9bef52d39R57-R71 ?

File: tasks/modules.py (hunk 2 above)
Comment (mackjmr) [reply tasks/modules.py:348]: See: https://dd.slack.com/archives/C06PBHLD4DQ/p1721057665355079

File: comp/otelcol/configstore/impl/configstore.go (hunk 3 above)
Comment (mackjmr) [reply comp/otelcol/configstore/impl/configstore.go:31]: Not sure to understand what you mean here ?

**Review by purple4reina — APPROVED** [review]

File: comp/otelcol/configstore/impl/configstore.go (hunk 3 above)
Comment (dineshg13) [reply comp/otelcol/configstore/impl/configstore.go:31]: On further thought, we can ignore this for now. We will revisit this later

**Review by dineshg13 — APPROVED** [review]

**Pull request merged**